### ISSUE TITLE
Run doc tests on README

### ISSUE THREAD

**Issue by huonw** [rust-lang/cargo#383]
`cargo test` could run (the equivalent of) `rustdoc --test README.md`, since people often have a basic example in it.

See also
- #739
- [doc-comment](https://crates.io/crates/doc-comment)
- [cargo-readme](https://crates.io/crates/cargo-readme)

Past efforts
- #445

**Comment by skade** [comment]
I'd like to push this one, I just ran into precisely the problem that my README examples were the only ones broken.

**Comment by pzol** [comment]
+1

**Comment by pythonesque** [comment]
I would love this.

**Comment by moonglum** [comment]
:+1:

**Comment by s-panferov** [comment]
Always forget to update README.md in my projects :+1:

**Comment by huonw** [comment]
cc https://github.com/brson/rust-skeptic

**Comment by carols10cents** [comment]
potentially relevant: https://github.com/rust-lang/rfcs/pull/1990

**Comment by abonander** [comment]
The external docs include (RFC 1990) is definitely the best way to go about this, IMO. You could include your README.md as the docs for a hidden item and then `rustdoc` should be able to automatically test the examples. There's a bit of engineering effort left to make it stabilizable though.

**Comment by abonander** [comment]
In case anyone missed it, I've got a proof-of-concept project for using the `external_doc` feature to test code examples in README.md: https://github.com/abonander/readme-doctest-poc

This can be done conditionally under a Cargo feature so it doesn't require nightly to test; once the feature is stabilized I think we can call this issue closed.

cc https://github.com/rust-lang/rust/issues/44732#issuecomment-456673619

**Comment by joshtriplett** [comment]
I'd like to revive this. I've seen projects do CI on their README by manually running `rustdoc test README.md` and specifying their crate as an `--extern`.

**Comment by shepmaster** [comment]
I've been happily using [doc-comment](https://crates.io/crates/doc-comment) to do this.

**Comment by ibraheemdev** [comment]
For anyone who doesn't want to add dependencies, this little macro works:
```rust

**Comment by tv42** [comment]
Three years later, here's an update to https://github.com/rust-lang/cargo/issues/383#issuecomment-720873790:

For anyone who doesn't want to add another dependency, this works:

```rust
#[cfg(doctest)]
#[doc = include_str!("../README.md")]
pub struct ReadMe;
```

Note, it can't be in `#[cfg(test)]`, it can't be in `tests/`.

Your IDE may gray it out as code not being used, because of the `#[cfg(doctest)]`. You may prefer

```rust
#[doc = include_str!("../README.md")]
struct _ReadMe;
```

Improvements welcome.

**Comment by tv42** [comment]
And here's the official documentation for it: https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html#include-items-only-when-collecting-doctests

Considering that many README.md's likely contain broken code and not-even-intended-to-be-complete examples, is this amount of opt-in enough; can this issue be closed?

**Comment by BilakshanP** [comment]
Hi, I didn't knew that this is an intended feature and by-chance found a bug. (unsure)

![image](https://github.com/rust-lang/cargo/assets/70910500/940faf5e-6801-468f-9be8-319a2dec02f8)

![image](https://github.com/rust-lang/cargo/assets/70910500/311753a7-6948-4de8-807f-dc460ed2db26)

As here you can see, I ran `cargo test` instead of `cargo test --lib`. All of my unit tests passed, but I found that some `doctest` was failing. Which is strange as I couldn't recall writing any doctests. As suggested by the output, I noticed that my `line 56` in lib.rs is error-prone. Which was quite strange for me as `lib.rs` is `< 30 LOC`. So, I ended up searching for that error-prone piece of code (as you can see in the right of the first image).

TL;DR + Conclusion: So, I think it is wrongly reporting that the issue is in `lib.rs`'s `line 56` when it is actually in `README`'s `line 56`

Please, let me know if I am somehow mistaken and this is the expected behavior.

**Comment by epage** [comment]
Its hard to tell whats going on without a reproduction case.  I'm assuming you are including your README in the documentation for your lib.  I have also seen line numbers be a bit strange in these scenarios.  I believe rustdoc is what determines all of that.

**Comment by BilakshanP** [comment]
> Its hard to tell whats going on without a reproduction case. I'm assuming you are including your README in the documentation for your lib. I have also seen line numbers be a bit strange in these scenarios. I believe rustdoc is what determines all of that.

If you don't mind may I share the link to the repository? So, that you can reproduce it yourselves.

**Comment by Marcono1234** [comment]
Even though you can now easily include the README and its code blocks as doctests (see https://github.com/rust-lang/cargo/issues/383#issuecomment-1832900297 above), two remaining inconveniences are:

- specifying attributes

   This depends on where / how the README is rendered. It seems GitHub supports the same (or a similar) syntax as [rustdoc](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html#attributes) like ```` ```rust,<other-attributes> ````. But the [CommonMark spec](https://spec.commonmark.org/0.31.2/#info-string) "does not mandate any particular treatment of the info string". So other Markdown renderers might not support specifying additional attributes this way. (Maybe not an issue if all common Markdown renderers support it.)

- hiding boilerplate lines

   There is no equivalent to [rustdoc's `# ` prefix](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html#hiding-portions-of-the-example) for regular Markdown code blocks, which means the code blocks have to include redundant boilerplate code to be compilable (especially a problem when using the `?` operator).

You can work around this to some extend by defining a custom proc-macro which transforms the Markdown code blocks in advance, but that seems to have several drawbacks like less useful test result output (does not point to the original Markdown file anymore, line numbers don't match).

Not sure if this would be a good idea, but could rustdoc recognize special HTML comments (something like `<!-- rustdoc-insert: Ok::<(), io::Error>(()) -->`) immediately in front of and behind the Markdown code blocks, which could then permit specifying rustdoc attributes and inserting additional (hidden) Rust code at the start / end of a code block?
